Prepare Verus-related crates for posting to crates.io

parno opened the pull request:
This is a first step in the plan discussed in #1790.  Specifically, we
1. Rename our various crates (except `vstd`) to include a `verus` specifier, to be polite in our usage of the crates.io namespace. 
2. Adjust the syntax macro so that all references to `builtin` and `builtin_macros` go through the existing re-export in `vstd::prelude`.  This way, clients only need to list `vstd` as an explicit dependency in their `Cargo.toml` file.
3. Add/adjust meta-data in our various `Cargo.toml` files for crates.io purposes.

Step 1 touches a lot of files in uninteresting ways, so it may be easier to review this PR commit-by-commit, rather than using the final diff.  If the bulk renaming commits need special handling from git, let me know how that's done.

The next steps include:
1. Finding a `synstructure` version we can use (instead of a git commit hash), or removing our dependency on `synstructure`
2. Developing a script to keep our package versions numbers and dependencies in sync.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/verus-lang/verus/blob/main/LICENSE).</small>

parno commented:
Also of interest to @jaybosamiya-ms

parno commented:
@tjhance The failing test is in `source/rust_verify_test/tests/basic.rs`, named `test_is_core`.

tjhance commented:
in fact, does `--no-vstd` still work at all?

parno commented:
> how does this effect tests? right now most tests don't import vstd for the sake of efficiency. Is that still the case?

This diff changed how we refer to builtin inside the syntax macro.
```
+++ b/source/builtin_macros/src/syntax.rs
@@ -5114,9 +5114,11 @@ pub(crate) struct Builtin(pub Span);
 impl ToTokens for Builtin {
     fn to_tokens(&self, tokens: &mut TokenStream) {
         if crate::cfg_verify_core() {
-            tokens.extend(quote_spanned! { self.0 => crate::verus_builtin });
+            tokens.extend(quote_spanned! { self.0 => crate::prelude });
+        } else if crate::cfg_verify_vstd() {
+            tokens.extend(quote_spanned! { self.0 => crate::prelude });
         } else {
-            tokens.extend(quote_spanned! { self.0 => ::verus_builtin });
+            tokens.extend(quote_spanned! { self.0 => ::vstd::prelude });
         }
     }
 }
```

This then prompted a change to `source/rust_verify_test/tests/common/mod.rs`, where I added `vstd::prelude` here:
```
 pub const USE_PRELUDE: &str = crate::common::code_str! {
+    use vstd::prelude::*;
     use verus_builtin::*;
     use verus_builtin_macros::*;
 };
```

Chris-Hawblitzel commented on source/builtin/Cargo.toml:
```diff
@@ -2,6 +2,14 @@
 version = "0.1.0"
 edition = "2018"
+license = "MIT"
+description = "Verus Builtins: Rust definitions for core constructs in the Verus language"
```
I would say "in Verus" rather than "in the Verus language"

parno replied:
Done.

jaybosamiya-ms commented on source/builtin/Cargo.toml:
```diff
@@ -2,6 +2,14 @@
+documentation = "https://verus-lang.github.io/verus/verusdoc/vstd/pervasive"
+homepage = "https://github.com/verus-lang/verus"
+repository = "https://github.com/verus-lang/verus"
+readme = "../../README.md"
```
Do we want to have a README specific to `builtin`/`builtin_macros` that points people at Verus and/or vstd instead?  Might be a little more convenient for folks who end up looking at these on crates.io to just directly get redirected.

parno replied:
Do you mean we should add a tiny README.md for `builtin`/`builtin_macros` that says something to the effect of "Support crates for Verus.  See https://github.com/verus-lang/verus for more details"?

jaybosamiya-ms replied:
Yeah, that's what I had in mind.  Having thought about it a bit, I don't know if that is really necessary though, since the way the current README is set up, they are told that this is Verus related, and adding a redirect that points them at literally that same README seems unnecessary to do until we see that someone was genuinely confused.

parno replied:
Okay, sounds good.  Thanks for helping us think through this process!

jaybosamiya-ms commented on source/builtin_macros/Cargo.toml:
```diff
@@ -18,8 +18,8 @@ proc-macro2 = "1.0.39"
-syn_verus = { path="../../dependencies/syn", features = ["full", "visit", "visit-mut", "extra-traits"] }
-prettyplease_verus = { path="../../dependencies/prettyplease" }
+syn_verus = { version = "2.0.96", path="../../dependencies/syn", features = ["full", "visit", "visit-mut", "extra-traits"] }
+prettyplease_verus = { version = "0.2.29", path="../../dependencies/prettyplease" }
```
Should these be `verus_syn` and `verus_prettyplease` (for the namespacing consistency)?

jaybosamiya-ms commented:
I added some comments inline, but also some more minor points:
- I think at some point in the past, there was a (brief) mention of `builtin` should probably have been called `builtins` but now we were stuck with `builtin`; if we are renaming to `verus_builtin`, this is probably the perfect time to update that if we still want to update that.  I personally think `builtins_macros` is too awkward for that, and having a `builtin_macros` alongside a `builtins` seems mildly annoying inconsistency too.  So yeah I don't think we should really use `builtins` but nonetheless, if we want to consider it, now is the time.
- When specifying version numbers for the dependencies of vstd (e.g., builtin), we might want to do an explicit `=` version requirement, otherwise, the defaults are ["caret" version requirements](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#caret-requirements).  Generally using caret version requirements _is_ the right move, since it allows semver-compatible updates to happen.  However, we are very _very_ particular about versions being in sync, in which case, we should be limiting it to exact version matches, rather than a caret requirement which means "the latest version that is semver-compatible with whatever was specified here".

parno commented:
Another oddity is that with the current version, I can successfully build Verus and verify vstd with vargo.  In a test Verus project that only lists `vstd` as a dependency, `cargo verus verify` succeeds, but `cargo build` fails when trying to compile `vstd`, reporting `failed to resolve: could not find vstd in the list of imported crates` in three spots:
```
   --> /Users/parno/git/verus/source/vstd/cell.rs:367:40
    |
367 |     pub fn new(val: T, Ghost(f): Ghost<spec_fn(T) -> bool>) -> (cell: Self)
    |                                        ^^^^^^^ could not find `vstd` in the list of imported crates

  --> /Users/parno/git/verus/source/vstd/map.rs:33:14
   |
33 |     mapping: spec_fn(K) -> Option<V>,
   |              ^^^^^^^ could not find `vstd` in the list of imported crates

   --> /Users/parno/git/verus/source/vstd/rwlock.rs:242:32
    |
242 | impl<V> RwLockPredicate<V> for spec_fn(V) -> bool {
    |                                ^^^^^^^ could not find `vstd` in the list of imported crates
```
which appear to be three places where, without `verus_keep_ghost`, the `::vstd::prelude::FnSpec` name "leaks" to the compiler.

parno commented:
@tjhance I think I've addressed your concern.  The `test_is_core` still fails with ~148 errors of the form:
```
error[E0433]: failed to resolve: unresolved import
    --> /Users/parno/git/verus/source/target/debug/test_inputs/basic-d6217e7fe140fc8d-test_is_core/../../../../vstd/seq_lib.rs:3562:17
     |
803  | /             assert_seqs_equal!(
804  | |                 self.subrange(k, self.len() as int).drop_last() ==
805  | |                 self.subrange(k, self.len() - 1)
806  | |             );
     | |_____________- in this macro invocation
...
3562 |           $crate::prelude::verus_proof_macro_exprs!($crate::vstd::seq_lib::assert_seqs_equal_internal!($($tail)*))
     |                   ^^^^^^^ unresolved import
```
I think this is because I changed some of the paths used in the macros written in `vstd`, e.g., 
```
macro_rules! assert_seqs_equal {
    [$($tail:tt)*] => {
-        ::builtin_macros::verus_proof_macro_exprs!($crate::vstd::seq_lib::assert_seqs_equal_internal!($($tail)*))
+        $crate::prelude::verus_proof_macro_exprs!($crate::vstd::seq_lib::assert_seqs_equal_internal!($($tail)*))
    };
}
```
Is there a way to write such macros that's generic as to whether we're in the i`s-core` mode?
